botocore/utils.py

```diff
@@ -243,28 +238,108 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
     source[current_key] = value
 
 
+class _RetriesExceededError(Exception):
+    """Internal exception used when the number of retries are exceeded."""
+    pass
+
+
 class InstanceMetadataFetcher(object):
 
-    _REQUIRED_CREDENTIAL_FIELDS = [
-        'AccessKeyId', 'SecretAccessKey', 'Token', 'Expiration'
-    ]
+    _RETRIES_EXCEEDED_ERROR_CLS = _RetriesExceededError
 
     def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
-                 num_attempts=1, url=METADATA_SECURITY_CREDENTIALS_URL,
-                 env=None, user_agent=None):
+                 num_attempts=1, base_url=METADATA_BASE_URL,
+                  env=None, user_agent=None):
         self._timeout = timeout
         self._num_attempts = num_attempts
-        self._url = url
+        self._base_url = base_url
         if env is None:
             env = os.environ.copy()
         self._disabled = env.get('AWS_EC2_METADATA_DISABLED', 'false').lower()
         self._disabled = self._disabled == 'true'
         self._user_agent = user_agent
         self._session = botocore.httpsession.URLLib3Session(
             timeout=self._timeout,
-            proxies=get_environ_proxies(self._url),
+            proxies=get_environ_proxies(self._base_url),
         )
 
+    def get_request(self, url_path, retry_func=None):
```

Comment:
So I didn't mind this as an internal method on the original imds fetcher class, but this just seems like a weird public interface to me (specifically the `retry_func` part).  Is there any way we can just keep the retry stuff in the original class that fetches iam role credentials?  This class could just return the response and the class that's retrieving the iam role could still decide if it needed to retry.

Reply:
This was to make it easier to write more clients since they all share that functionality. Making it public was to indicate that its a publicly facing function since it will have a consumer in a different project (CLI).